### PR TITLE
Improve docs

### LINKED ISSUE #802: improve autogenerated documentation
the autogenerated documentation has much room for improvement. In detail:

- no manual update should be necessary, as this documentation is supposed to be autogenerated
- current errors in building the docs should be fixed
- the python code should be documented (currently it's not)

### PR DESCRIPTION
**Pull request by mgiulini**
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and that you comply with the following criteria:

- [ ] You have sticked to Python. Please talk to us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [ ] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [ ] You structured the code into small functions as much as possible. You can use classes if there is a (state) purpose
- [ ] Your code follows our coding style
- [ ] You wrote tests for the new code
- [ ] `tox` tests pass. *Run `tox` command inside the repository folder*
- [ ] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [ ] PR does not add any dependencies, unless permission granted by the HADDOCK team
- [ ] PR does not break licensing
- [x] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

Closes #802 . This PR is supposed to substantially improve the autogenerated documentation created when running `tox -e docs`. More specifically:

- the documentation is now searchable (thanks @rvhonorato )
- the bugs that were stopping the build of the documentation have been corrected
- python code docs are now properly rendered
- sphinx-apidoc runs before sphinx-build to create the rst files
- the rst files are almost completely removed from the repo. The old rst files were outdated and basically their presence forced the developer to change them at every commit. Not exactly desirable for an autogenerated documentation
- the devtools script `buiild_defaults_rsts` is adapted to move and prettify the rst files in the `docs` folder

### CODE REVIEW

**Review by rvhonorato — CHANGES_REQUESTED** [review]
Nice!

When you run the sphinx commands it generates a lot of these `.rst` files that are not in the `.gitignore`, so I'm guessing you edit it or should they be in the repository?

```
On branch improve_docs
Your branch is up to date with 'origin/improve_docs'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	docs/haddock.clis.cli.rst
	docs/haddock.clis.cli_analyse.rst
	docs/haddock.clis.cli_bm.rst
	docs/haddock.clis.cli_cfg.rst
	docs/haddock.clis.cli_clean.rst
	docs/haddock.clis.cli_cp.rst
	docs/haddock.clis.cli_dmn.rst
	docs/haddock.clis.cli_mpi.rst
	docs/haddock.clis.cli_pp.rst
	docs/haddock.clis.cli_re.rst
	docs/haddock.clis.cli_restraints.rst
	docs/haddock.clis.cli_score.rst
	docs/haddock.clis.cli_traceback.rst
	docs/haddock.clis.cli_unpack.rst
	docs/haddock.clis.re.clustfcc.rst
	docs/haddock.clis.re.clustrmsd.rst
	docs/haddock.clis.re.rst
	docs/haddock.clis.re.score.rst
	docs/haddock.clis.restraints.active_passive_to_ambig.rst
	docs/haddock.clis.restraints.passive_from_active.rst
	docs/haddock.clis.restraints.restrain_bodies.rst
	docs/haddock.clis.restraints.rst
	docs/haddock.clis.restraints.validate_tbl.rst
	docs/haddock.clis.rst
	docs/haddock.core.cns_paths.rst
	docs/haddock.core.defaults.rst
	docs/haddock.core.exceptions.rst
	docs/haddock.core.rst
	docs/haddock.core.supported_molecules.rst
	docs/haddock.core.typing.rst
	docs/haddock.gear.clean_steps.rst
	docs/haddock.gear.config.rst
	docs/haddock.gear.expandable_parameters.rst
	docs/haddock.gear.extend_run.rst
	docs/haddock.gear.greetings.rst
	docs/haddock.gear.haddockmodel.rst
	docs/haddock.gear.parameters.rst
	docs/haddock.gear.prepare_run.rst
	docs/haddock.gear.preprocessing.rst
	docs/haddock.gear.restart_run.rst
	docs/haddock.gear.rst
	docs/haddock.gear.validations.rst
	docs/haddock.gear.yaml2cfg.rst
	docs/haddock.gear.zerofill.rst
	docs/haddock.libs.libalign.rst
	docs/haddock.libs.libcli.rst
	docs/haddock.libs.libclust.rst
	docs/haddock.libs.libcns.rst
	docs/haddock.libs.libfunc.rst
	docs/haddock.libs.libhpc.rst
	docs/haddock.libs.libinteractive.rst
	docs/haddock.libs.libio.rst
	docs/haddock.libs.liblog.rst
	docs/haddock.libs.libmath.rst
	docs/haddock.libs.libmpi.rst
	docs/haddock.libs.libontology.rst
	docs/haddock.libs.libparallel.rst
	docs/haddock.libs.libpdb.rst
	docs/haddock.libs.libplots.rst
	docs/haddock.libs.librestraints.rst
	docs/haddock.libs.libstructure.rst
	docs/haddock.libs.libsubprocess.rst
	docs/haddock.libs.libtimer.rst
	docs/haddock.libs.libutil.rst
	docs/haddock.libs.libworkflow.rst
	docs/haddock.libs.rst
	docs/haddock.modules.analysis.alascan.rst
	docs/haddock.modules.analysis.alascan.scan.rst
	docs/haddock.modules.analysis.caprieval.capri.rst
	docs/haddock.modules.analysis.caprieval.rst
	docs/haddock.modules.analysis.clustfcc.clustfcc.rst
	docs/haddock.modules.analysis.clustfcc.rst
	docs/haddock.modules.analysis.clustrmsd.clustrmsd.rst
	docs/haddock.modules.analysis.clustrmsd.rst
	docs/haddock.modules.analysis.contactmap.contmap.rst
	docs/haddock.modules.analysis.contactmap.rst
	docs/haddock.modules.analysis.ilrmsdmatrix.ilrmsd.rst
	docs/haddock.modules.analysis.ilrmsdmatrix.rst
	docs/haddock.modules.analysis.rmsdmatrix.rmsd.rst
	docs/haddock.modules.analysis.rmsdmatrix.rst
	docs/haddock.modules.analysis.rst
	docs/haddock.modules.analysis.seletop.rst
	docs/haddock.modules.analysis.seletopclusts.rst
	docs/haddock.modules.extras.exit.rst
	docs/haddock.modules.extras.rst
	docs/haddock.modules.refinement.emref.rst
	docs/haddock.modules.refinement.flexref.rst
	docs/haddock.modules.refinement.mdref.rst
	docs/haddock.modules.refinement.rst
	docs/haddock.modules.sampling.gdock.rst
	docs/haddock.modules.sampling.lightdock.rst
	docs/haddock.modules.sampling.rigidbody.rst
	docs/haddock.modules.sampling.rst
	docs/haddock.modules.scoring.emscoring.rst
	docs/haddock.modules.scoring.mdscoring.rst
	docs/haddock.modules.scoring.rst
	docs/haddock.modules.topology.rst
	docs/haddock.modules.topology.topoaa.rst
	docs/haddock.modules.topology.topocg.rst

nothing added to commit but untracked files present (use "git add" to track)
```

Other than that it looks functional to me.

File: docs/conf.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import subprocess
 from pathlib import Path
```
Comment (rvhonorato) [inline docs/conf.py:9]: I see the docs show "Copyright 2022" - you should be able to set it dynamically using `datetime` here;

```python
import datetime 
# ...
year = datetime.date.today().year
# ...
```

**Comment by mgiulini** [comment]
> Nice!
> 
> When you run the sphinx commands it generates a lot of these `.rst` files that are not in the `.gitignore`, so I'm guessing you edit it or should they be in the repository?
> 
> ```
> On branch improve_docs
> Your branch is up to date with 'origin/improve_docs'.
> 
> Untracked files:
>   (use "git add <file>..." to include in what will be committed)
> 	docs/haddock.clis.cli.rst
> 	docs/haddock.clis.cli_analyse.rst
> 	docs/haddock.clis.cli_bm.rst
> 	docs/haddock.clis.cli_cfg.rst
> 	docs/haddock.clis.cli_clean.rst
> 	docs/haddock.clis.cli_cp.rst
> 	docs/haddock.clis.cli_dmn.rst
> 	docs/haddock.clis.cli_mpi.rst
> 	docs/haddock.clis.cli_pp.rst
> 	docs/haddock.clis.cli_re.rst
> 	docs/haddock.clis.cli_restraints.rst
> 	docs/haddock.clis.cli_score.rst
> 	docs/haddock.clis.cli_traceback.rst
> 	docs/haddock.clis.cli_unpack.rst
> 	docs/haddock.clis.re.clustfcc.rst
> 	docs/haddock.clis.re.clustrmsd.rst
> 	docs/haddock.clis.re.rst
> 	docs/haddock.clis.re.score.rst
> 	docs/haddock.clis.restraints.active_passive_to_ambig.rst
> 	docs/haddock.clis.restraints.passive_from_active.rst
> 	docs/haddock.clis.restraints.restrain_bodies.rst
> 	docs/haddock.clis.restraints.rst
> 	docs/haddock.clis.restraints.validate_tbl.rst
> 	docs/haddock.clis.rst
> 	docs/haddock.core.cns_paths.rst
> 	docs/haddock.core.defaults.rst
> 	docs/haddock.core.exceptions.rst
> 	docs/haddock.core.rst
> 	docs/haddock.core.supported_molecules.rst
> 	docs/haddock.core.typing.rst
> 	docs/haddock.gear.clean_steps.rst
> 	docs/haddock.gear.config.rst
> 	docs/haddock.gear.expandable_parameters.rst
> 	docs/haddock.gear.extend_run.rst
> 	docs/haddock.gear.greetings.rst
> 	docs/haddock.gear.haddockmodel.rst
> 	docs/haddock.gear.parameters.rst
> 	docs/haddock.gear.prepare_run.rst
> 	docs/haddock.gear.preprocessing.rst
> 	docs/haddock.gear.restart_run.rst
> 	docs/haddock.gear.rst
> 	docs/haddock.gear.validations.rst
> 	docs/haddock.gear.yaml2cfg.rst
> 	docs/haddock.gear.zerofill.rst
> 	docs/haddock.libs.libalign.rst
> 	docs/haddock.libs.libcli.rst
> 	docs/haddock.libs.libclust.rst
> 	docs/haddock.libs.libcns.rst
> 	docs/haddock.libs.libfunc.rst
> 	docs/haddock.libs.libhpc.rst
> 	docs/haddock.libs.libinteractive.rst
> 	docs/haddock.libs.libio.rst
> 	docs/haddock.libs.liblog.rst
> 	docs/haddock.libs.libmath.rst
> 	docs/haddock.libs.libmpi.rst
> 	docs/haddock.libs.libontology.rst
> 	docs/haddock.libs.libparallel.rst
> 	docs/haddock.libs.libpdb.rst
> 	docs/haddock.libs.libplots.rst
> 	docs/haddock.libs.librestraints.rst
> 	docs/haddock.libs.libstructure.rst
> 	docs/haddock.libs.libsubprocess.rst
> 	docs/haddock.libs.libtimer.rst
> 	docs/haddock.libs.libutil.rst
> 	docs/haddock.libs.libworkflow.rst
> 	docs/haddock.libs.rst
> 	docs/haddock.modules.analysis.alascan.rst
> 	docs/haddock.modules.analysis.alascan.scan.rst
> 	docs/haddock.modules.analysis.caprieval.capri.rst
> 	docs/haddock.modules.analysis.caprieval.rst
> 	docs/haddock.modules.analysis.clustfcc.clustfcc.rst
> 	docs/haddock.modules.analysis.clustfcc.rst
> 	docs/haddock.modules.analysis.clustrmsd.clustrmsd.rst
> 	docs/haddock.modules.analysis.clustrmsd.rst
> 	docs/haddock.modules.analysis.contactmap.contmap.rst
> 	docs/haddock.modules.analysis.contactmap.rst
> 	docs/haddock.modules.analysis.ilrmsdmatrix.ilrmsd.rst
> 	docs/haddock.modules.analysis.ilrmsdmatrix.rst
> 	docs/haddock.modules.analysis.rmsdmatrix.rmsd.rst
> 	docs/haddock.modules.analysis.rmsdmatrix.rst
> 	docs/haddock.modules.analysis.rst
> 	docs/haddock.modules.analysis.seletop.rst
> 	docs/haddock.modules.analysis.seletopclusts.rst
> 	docs/haddock.modules.extras.exit.rst
> 	docs/haddock.modules.extras.rst
> 	docs/haddock.modules.refinement.emref.rst
> 	docs/haddock.modules.refinement.flexref.rst
> 	docs/haddock.modules.refinement.mdref.rst
> 	docs/haddock.modules.refinement.rst
> 	docs/haddock.modules.sampling.gdock.rst
> 	docs/haddock.modules.sampling.lightdock.rst
> 	docs/haddock.modules.sampling.rigidbody.rst
> 	docs/haddock.modules.sampling.rst
> 	docs/haddock.modules.scoring.emscoring.rst
> 	docs/haddock.modules.scoring.mdscoring.rst
> 	docs/haddock.modules.scoring.rst
> 	docs/haddock.modules.topology.rst
> 	docs/haddock.modules.topology.topoaa.rst
> 	docs/haddock.modules.topology.topocg.rst
> 
> nothing added to commit but untracked files present (use "git add" to track)
> ```
> 
> Other than that it looks functional to me.

basically the configuration file calls this `build_default_rst` script that manages these rst files..I guess you just ran the sphinx commands, correct? try to run `tox -e docs` and these untracked files should disappear

File: docs/conf.py (hunk 1 above)
Comment (mgiulini) [reply docs/conf.py:9]: good point!

**Comment by rvhonorato** [comment]
those are generated after `tox -e docs` shouldn't they be in `.gitignore`?

**Pull request merged**